Support RealVectorValue in BatchMeshGenerationAction

miaoyinb opened the pull request:
closes #30910

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2998490/), step Docs: sync website on 8e8ec05 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30911/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2998539/), step Generate coverage on 8e8ec05 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>98f5d7</th>
      <th colspan=4>#30911 8e8ec0</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.47%</td>
      <td>85.47%</td>
      <td>+0.00%</td>
      <td>92.59%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>114883</td>
      <td>114899</td>
      <td>+16</td>
      <td>25</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19525</td>
      <td>19527</td>
      <td>+2</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30911/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30911/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30911/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3003239/), step Clang format on 7253535 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30911/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30911/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0f68102779a6ca115fa48d28601327e4c3133dee`

miaoyinb commented on framework/src/actions/BatchMeshGeneratorAction.C:
```diff
@@ -631,19 +652,32 @@ BatchMeshGeneratorAction::convertAndSetStringLikeVector(
+                                                    const std::vector<std::string> & param_names,
+                                                    const std::vector<ParameterType> & param_types)
+{
+  std::vector<bool> is_real_vector_value(param_names.size(), false);
```
Used to record whether a `vector<Real>` is actually a `RealVectorValue`

lindsayad replied:
This one is just a dummy vector though right? Because this overload is called for the scalar parameters?

https://github.com/idaholab/moose/blob/2bbbe68aee31f3de438f0897047f241f733d8e91/framework/src/actions/BatchMeshGeneratorAction.C#L198-L206

I think this dummy vector is what confused me

miaoyinb replied:
Yes, when the method is used to check scalar type parameters, this vector is dummy.

miaoyinb replied:
Any suggestion to further optimize this?

miaoyinb replied:
I am resolving this as the code has been redesigned and needs to be re-reviewed.

lindsayad commented on framework/src/actions/BatchMeshGeneratorAction.C:
```diff
@@ -631,19 +652,32 @@ BatchMeshGeneratorAction::convertAndSetStringLikeVector(
     {
       case (ParameterType::REAL):
-        checkInputParameterType<Real>(params, action_input_param_name, param_names[i], is_vector);
+        is_real_vector_value[i] = checkInputParameterType<Real, RealVectorValue>(
+            params, action_input_param_name, param_names[i], is_vector);
```
I searched through the code again and I still don't see where you use the vector version of `is_real_vector_value`

miaoyinb replied:
Each element of this vector is used when I call `setVectorParams()`, which is the fourth argument.

For example,
```
      setVectorParams(params,
                      _fixed_vector_input_param_names[j],
                      _fixed_vector_input_param_types[j],
                      _fixed_vector_input_param_is_real_vector_value[j],
                      _fixed_vector_input_param_values[j]);
```

miaoyinb replied:
`is_real_vector_value` is a reference argument of `checkInputParametersTypes()`

`checkInputParametersTypes()` is called twice in this C file with this argument specified.

Line 212, with `_batch_vector_input_param_is_real_vector_value` provided as this argument
Line 221, with `_fixed_vector_input_param_is_real_vector_value` provided as this argument

Both `_batch_vector_input_param_is_real_vector_value` and `_fixed_vector_input_param_is_real_vector_value` are then used as an argument in `setVectorParams()` (Line 474 and Line 485).

lindsayad reviewed (changes requested):
I think this should be re-designed a bit. I think that you should add a new enumerator under the `ParameterType` enum for `REALVECTORVALUE`. And then I think you should be able to support mesh generator parameters like `RealVectorValue` under your `scalar` action parameters and (hypothetical) mesh generator parameters like `std::vector<RealVectorValue>` under your `vector` action parameters. Then `RealVectorValue` and e.g. `Real` parameter types are conceptually treated the same. 

Because what happens if in the future someone wants to do a `RealTensorValue` mesh generator parameter? Would you also put support for that in the `vector` action params?

Even though we made `checkInputParameterTypes` semantically `const` they don't feel conceptually `const` because we are now using calls to these methods to modify data members. To me `checkInputParameterTypes` should only be a check. It shouldn't be responsible for actually determining/setting parameter types

miaoyinb commented:
`REALVECTORVALUE` is a vector consisting of three real elements. It serves as a vector input parameter, rather than a scalar one, and therefore is not syntactically compatible with other scalar-type input parameters. Would it be appropriate to classify `REALVECTORVALUE` under `vector` types instead of `scalar` types?

miaoyinb commented:
It would still need some work to make it work with other scalar input in one input parameter. I will think about it.

lindsayad commented:
I just want to make sure we're thinking about the future. I just think it would be harder to maintain if we need to add a new method parameter to `checkParameterTypes` every time we want to add a new compound type
```c++
  void checkInputParametersTypes(const InputParameters & params,
                                 const std::string & action_input_param_name,
                                 const std::vector<std::string> & param_names,
                                 const std::vector<ParameterType> & param_types,
                                 const bool & is_vector,
                                 std::vector<bool> & is_vector_real_vector_value,
                                 std::vector<bool> & is_vector_real_tensor_value,
                                 std::vector<bool> & is_vector_rank_two_tensor,
                                 std::vector<bool> & is_vector_rank_four_tensor) const;
```
so on and so forth

miaoyinb committed:
Support RealVectorValue in BatchMeshGenerationAction

closes #30910

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3014286/), step Clang format on be0ed2f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30911/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30911/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format bb6cbb87060f55c217d8a465b388bb067c58a3fa`

miaoyinb commented:
@lindsayad Just a friendly reminder. This has be redesigned and ready for another round of review. Thanks.

lindsayad committed:
Add iterator pair join function

lindsayad committed:
Use Moose::dim for clarity

And a tiny bit for flexibility although no one configures libMesh
when using MOOSE with a dimension other than 3

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3040164/), step Clang format on 107b730 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30911/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30911/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9ec918489917eb940c9fa8e2d7b43b74eb2a93ba`

lindsayad commented:
@miaoyinb see if you're ok with the last commit I pushed to use `Moose::dim` instead of 3
